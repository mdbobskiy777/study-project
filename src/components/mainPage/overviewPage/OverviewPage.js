import React from "react"
import styled from "styled-components"
import 'bootstrap/dist/css/bootstrap.min.css'
import {ListGroup} from "react-bootstrap";

const MainContainer = styled.div`
  text-align: center;
  margin: 0 auto;
  padding: 10px;
  max-width: 25%;
  background: whitesmoke;
`
const MyUL = styled(ListGroup)`
  margin: 10px auto;
  padding: 10px;
  background: white;
`
const MyLI = styled(ListGroup.Item)`
  width: 100%;
  margin: 0 auto;
`
const MyTittle = styled.div`
  background: white;
  margin: 5px auto;
  padding: 2px;
  font-size: 1.5em;
`
const MyDiv = styled.div`
  font-size: 1.25em;
`
const OverviewPage = () => {
    return (
        <MainContainer>
            <MyTittle>Leanna Hogg</MyTittle>
            <MyDiv>Direct subordinates:</MyDiv>
            <MyUL>
                <MyLI>Vincent Todd</MyLI>
                <MyLI>Faye Oneill</MyLI>
                <MyLI>Lynn Haigh</MyLI>
                <MyLI>Aila Hodgson</MyLI>
            </MyUL>
        </MainContainer>
    )
}
export default OverviewPage
