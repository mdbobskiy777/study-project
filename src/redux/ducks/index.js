import employers from './employers';
import searchedName from './searchedName';
import subordinates from './subordinates';
import {combineReducers} from 'redux';

export default combineReducers({
    employers:employers,
    searchedName:searchedName,
    subordinates:subordinates
});
