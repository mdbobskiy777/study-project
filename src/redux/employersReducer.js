import {call, put, takeEvery} from 'redux-saga/effects'

import {EmployersAPI} from "../api/api";

export const SET_EMPLOYERS = "SET_EMPLOYERS"
export const SET_SUBORDINATES = "GET_SUBORDINATES"
export const SET_SEARCHED_NAME = "SET_SEARCHED_NAME"
export const REQUESTED_EMPLOYERS_FAILED = "REQUESTED_EMPLOYERS_FAILED"
export const REQUESTED_SUBORDINATES_FAILED = "REQUESTED_SUBORDINATES_FAILED"
export const FETCHED_EMPLOYERS = "FETCHED_EMPLOYERS"
export const FETCHED_SUBORDINATES = "FETCHED_SUBORDINATES"

const initialState = {
    searchedName: '',
    employers: [],
    employerData: []
}
const employersReducer = (state = initialState, action) => {
    switch (action.type) {
        case SET_EMPLOYERS :
            return {
                ...state, employers: action.employers
            }
        case SET_SUBORDINATES:
            return {
                ...state,employerData: action.subordinatesData
            }
        case SET_SEARCHED_NAME:
            return {...state, searchedName: action.searchedName}
        default:
            return state
    }
}

//action creators
const setEmployersList = employers => ({type: SET_EMPLOYERS, employers})

export const setSearchedName = searchedName => ({type: SET_SEARCHED_NAME, searchedName})

const setSubordinatesList = subordinatesData => ({type: SET_SUBORDINATES, subordinatesData})

export const fetchEmployersList = () => {
    return {type: FETCHED_EMPLOYERS}

}
export const fetchSubordinates = (name) => {
    return {type: FETCHED_SUBORDINATES, name}

}
const requestEmployersError = () => {
    return {type: REQUESTED_EMPLOYERS_FAILED}
}
const requestSubordinatesError = () => {
    return {type: REQUESTED_SUBORDINATES_FAILED}
}

//sagas
export function* watchFetchEmployers() {
    yield takeEvery(FETCHED_EMPLOYERS, fetchEmployersAsync)
}

function* fetchEmployersAsync() {
    try {
        const employers = yield call(() => {
                return EmployersAPI.getEmployers()
            }
        )
        yield put(setEmployersList(employers))
    } catch (error) {
        yield put(requestEmployersError())
    }
}


export function* watchFetchSubordinates() {
    yield takeEvery(FETCHED_SUBORDINATES, fetchSubordinatesAsync)
}

function* fetchSubordinatesAsync(action) {

    try {
        debugger
        const subordinatesData = yield call(() => {
                return EmployersAPI.getSubordinates(action.name)
            }
        )
        yield put(setSubordinatesList(subordinatesData))
    } catch (error) {
        yield put(requestSubordinatesError())
    }
}


export default employersReducer
