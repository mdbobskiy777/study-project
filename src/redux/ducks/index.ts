import { combineReducers } from 'redux';

import employers from './employers';
import searchedName from './searchedName';
import subordinates from './subordinates';

export default combineReducers({
    employers: employers,
    searchedName: searchedName,
    subordinates: subordinates
});
