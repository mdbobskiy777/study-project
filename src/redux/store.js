import {applyMiddleware, combineReducers, compose, createStore} from "redux"
import employersReducer, {watchFetchEmployers, watchFetchSubordinates} from "./employersReducer"
import createSagaMiddleware from 'redux-saga'

const rootReducer = combineReducers({
    employersReducer: employersReducer
})
const sagaMiddleware = createSagaMiddleware()
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

const createAppStore = () => {
    const store = createStore(rootReducer,
        composeEnhancers(applyMiddleware(sagaMiddleware)))
    sagaMiddleware.run(watchFetchEmployers)
    sagaMiddleware.run(watchFetchSubordinates)
    return store
}


export const appStore = createAppStore();

