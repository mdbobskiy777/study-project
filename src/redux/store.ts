import { configureStore } from '@reduxjs/toolkit';
import createSagaMiddleware from 'redux-saga';

import employersReducer from './ducks';
import { watchFetchEmployers } from './ducks/employers';
import { watchFetchSubordinates } from './ducks/subordinates';

export type RootState = ReturnType<typeof appStore.getState>;
export type AppDispatch = typeof appStore.dispatch;

const rootReducer = employersReducer;

const sagaMiddleware = createSagaMiddleware();

const createAppStore = () => {
    const store = configureStore({
        reducer: rootReducer,
        middleware:[sagaMiddleware],
    });
    sagaMiddleware.run(watchFetchEmployers);
    sagaMiddleware.run(watchFetchSubordinates);

    return store;
};

export const appStore = createAppStore();

