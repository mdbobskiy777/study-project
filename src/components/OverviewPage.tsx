import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useParams } from 'react-router';
import { Col, Container, Row } from 'react-bootstrap';

import { fetchSubordinates, setFetching, setSubordinatesList } from '../redux/ducks/subordinates';
import OverviewPageStyled from '../styled/OverviewPageStyled';
import { AppDispatch, RootState } from '../redux/store';
import Preloader from '../assets/images/preloader.svg';

const OverviewPage = () : JSX.Element => {
    const employerDataSelector = useSelector((state:RootState) => state. subordinates.employerData);
    const isFetchingSelector = useSelector((state:RootState) => state. subordinates.isFetching);

    const dispatch = useDispatch<AppDispatch>();

    const paramsName: { name:string } = useParams();

    useEffect(() => {
        dispatch(setFetching(true));
        if (paramsName.name) dispatch(fetchSubordinates(paramsName.name));
    }, [paramsName.name]);

    useEffect(() => {
        return () => {
            dispatch(setSubordinatesList([]));
        };
    },[]);

    return (
        <Container>
            <OverviewPageStyled.MainContainer>
                <Row className="justify-content-center">
                    <Col xl="6" lg="6">
                        <OverviewPageStyled.MyTittle>{paramsName.name}</OverviewPageStyled.MyTittle>
                        {employerDataSelector[0] && (
                            <OverviewPageStyled.MyDiv>
                                {employerDataSelector[0] && (
                                    `Direct subordinates: ${employerDataSelector[0]}`
                                )}
                            </OverviewPageStyled.MyDiv>
                        )}
                    </Col>
                </Row>
                <Row className="justify-content-center">
                    <Col xl="6" lg="6">
                        <OverviewPageStyled.MyUL>
                            {employerDataSelector[1]?(
                                Object.values(employerDataSelector[1]).map(
                                    ( subordinate: string , index: React.Key ) =>
                                        <OverviewPageStyled.MyLI key={index}>
                                            {subordinate}
                                        </OverviewPageStyled.MyLI>
                                )) : (
                                !isFetchingSelector && (
                                    <div>Haven`t subordinates</div>
                                )
                            )}
                            {isFetchingSelector && (
                                <div>
                                    <OverviewPageStyled.MyIMG src={Preloader} />
                                </div>
                            )}
                        </OverviewPageStyled.MyUL>
                    </Col>
                </Row>
            </OverviewPageStyled.MainContainer>
        </Container>
    );
};

export default OverviewPage;
