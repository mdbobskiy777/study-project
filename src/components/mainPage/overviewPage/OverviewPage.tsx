import "bootstrap/dist/css/bootstrap.min.css";

import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router";
import { fetchSubordinates, setFetching, setSubordinatesList } from "../../../redux/ducks/subordinates";

import OverviewPageStyled from "../../../styled/mainPage/overviewPage/OverviewPageStyled";
import { Col, Container, Row } from "react-bootstrap";
import Preloader from "../../../assets/images/preloader.svg";
import {AppDispatch, RootState} from "../../../redux/store";

type ParamType = {
    name:string
};

const OverviewPage:React.FC = ():JSX.Element => {

    const employerDataSelector = useSelector((state:RootState) => state. subordinates.employerData);
    const isFetchingSelector = useSelector((state:RootState) => state. subordinates.isFetching);
    const dispatch = useDispatch<AppDispatch>();
    const paramsName:ParamType = useParams();

    useEffect(() => {
        dispatch(setFetching(true));
        if (paramsName.name) dispatch(fetchSubordinates(paramsName.name));
    },
    [paramsName.name]);

    useEffect(() => {
        return () => {
            dispatch(setSubordinatesList([]));
        };
    },[]);

    return (
        <Container>
            <OverviewPageStyled.MainContainer>
                <Row xl="12" lg="10" md="8" sm="8" xs="6">
                    <Col xl="12" lg="12" md="12" sm="12" xs="12">
                        <OverviewPageStyled.MyTittle>{paramsName.name}</OverviewPageStyled.MyTittle>
                        {
                            employerDataSelector[0]&&<OverviewPageStyled.MyDiv>
                                {(employerDataSelector[0])&&`Direct subordinates: ${employerDataSelector[0]}`}
                            </OverviewPageStyled.MyDiv>
                        }
                    </Col>
                </Row>
                <Row xl="12" lg="10" md="8" sm="8" xs="6">
                    <Col xl="12" lg="12" md="12" sm="12" xs="12">
                        <OverviewPageStyled.MyUL>
                            {(employerDataSelector[1])?(
                            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                            // @ts-ignore
                                employerDataSelector[1]["direct-subordinates"].map(
                                    ( subordinate: string , index: React.Key ) =>
                                        <OverviewPageStyled.MyLI key={index}>
                                            {subordinate}
                                        </OverviewPageStyled.MyLI>)):
                                !isFetchingSelector&&<div>Haven`t subordinates</div>
                            }
                            {
                                isFetchingSelector&&<div>
                                    <OverviewPageStyled.MyIMG src = {Preloader} />
                                </div>
                            }
                        </OverviewPageStyled.MyUL>
                    </Col>
                </Row>
            </OverviewPageStyled.MainContainer>
        </Container>
    );
};

export default OverviewPage;
