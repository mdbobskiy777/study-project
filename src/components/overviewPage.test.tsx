import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { render } from '@testing-library/react';
import { act as domAct } from 'react-dom/test-utils';

import { appStore } from '../redux/store';
import OverviewPage from './OverviewPage';

it('OverviewPage list showing preloader', () => {
    const utils = render(
        <BrowserRouter>
            <Provider store={appStore}>
                <OverviewPage />
            </Provider>
        </BrowserRouter>
    );

    const preloader = utils.getByRole('img');

    expect(preloader);
});

describe('overviewPage render', () => {
    let container: HTMLElement;

    it('should render empty page and then page with preloader', async ()  =>  {
        await domAct(async () => {
            const result = render(
                <BrowserRouter>
                    <Provider store={appStore}>
                        <OverviewPage />
                    </Provider>
                </BrowserRouter>);

            container = result.container;

            expect(container.innerHTML).toMatchSnapshot();
        });

        expect(container.innerHTML).toMatchSnapshot();
    });
});
