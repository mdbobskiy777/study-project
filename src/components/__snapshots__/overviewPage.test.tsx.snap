// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overviewPage render should render empty page and then page with preloader 1`] = `"<div class=\\"container\\"><div class=\\"sc-bdfBwQ iEjbxj\\"><div class=\\"justify-content-center row\\"><div class=\\"col-xl-6 col-lg-6\\"><div class=\\"sc-hKgILt hWsmEw\\"></div></div></div><div class=\\"justify-content-center row\\"><div class=\\"col-xl-6 col-lg-6\\"><div class=\\"sc-gsTCUz iOjICt list-group\\"></div></div></div></div></div>"`;

exports[`overviewPage render should render empty page and then page with preloader 2`] = `"<div class=\\"container\\"><div class=\\"sc-bdfBwQ iEjbxj\\"><div class=\\"justify-content-center row\\"><div class=\\"col-xl-6 col-lg-6\\"><div class=\\"sc-hKgILt hWsmEw\\"></div></div></div><div class=\\"justify-content-center row\\"><div class=\\"col-xl-6 col-lg-6\\"><div class=\\"sc-gsTCUz iOjICt list-group\\"><div><img src=\\"preloader.svg\\" class=\\"sc-jSgupP emQGDX\\"></div></div></div></div></div></div>"`;
